File: avl_tree/AVLTree.java
package avl_tree;

import java.util.ArrayDeque;

public class AVLTree<T extends Comparable<T>> {
    private class Node<E extends Comparable<E>> {
        private int height;
        private E data;
        Node<E> parent;
        Node<E> left;
        Node<E> right;

        Node(E data) {
            this(data, null);
        }

        Node(E data, Node<E> parent) {
            this(data, parent, null, null);
        }

        Node(E data, Node<E> parent, Node<E> left, Node<E> right) {
            this.data = data;
            this.parent = parent;
            this.left = left;
            this.right = right;
            this.height = 0;
        }

        int getHeight() {
            return height;
        }

        void updateHeight() {
            int leftHeight = (left != null) ? (left.getHeight()) : -1;
            int rightHeight = (right != null) ? (right.getHeight()) : -1;

            height = 1 + Math.max(leftHeight, rightHeight);
        }
    }

    Node<T> root;
    int size;

    AVLTree() {
        size = 0;
    }

    AVLTree(T data) {
        size = 1;
        root = new Node<T>(data);
    }

    public boolean isEmpty() {
        return root == null;
    }

    public int size() {
        return size;
    }

    int height(Node<T> node) {
        return (node != null) ? (node.getHeight()) : -1;
    }

    int balanceFactor(Node<T> node) {
        if (node == null) {
            return 0;
        }

        int leftHeight = height(node.left);
        int rightHeight = height(node.right);

        return leftHeight - rightHeight;
    }

    private Node<T> leftRotate(Node<T> node) {
        if (node.right == null) {
            throw new IllegalArgumentException("Cannot right rotate null.");
        }

        Node<T> rightChild = node.right;

        node.right = rightChild.left;
        if (rightChild.left != null) {
            rightChild.left.parent = node;
        }

        rightChild.left = node;
        rightChild.parent = node.parent;
        if (rightChild.parent != null) {
            if (node.parent.left == node) {
                node.parent.left = rightChild;
            } else {
                node.parent.right = rightChild;
            }
        }

        node.parent = rightChild;

        node.updateHeight();
        rightChild.updateHeight();

        return rightChild;
    }

    private Node<T> rightRotate(Node<T> node) {
        if (node.left == null) {
            throw new IllegalArgumentException("Cannot left rotate null.");
        }

        Node<T> leftChild = node.left;

        node.left = leftChild.right;

        if (leftChild.right != null) {
            leftChild.right.parent = node;
        }

        leftChild.right = node;
        leftChild.parent = node.parent;
        if (leftChild.parent != null) {
            if (node.parent.left == node) {
                node.parent.left = leftChild;
            } else {
                leftChild.parent.right = leftChild;
            }
        }

        node.parent = leftChild;

        node.updateHeight();
        leftChild.updateHeight();

        return leftChild;
    }

    private Node<T> rebalanceNode(Node<T> node) {
        if (balanceFactor(node) > 1) {
            if (balanceFactor(node.left) < 0) {
                node.left = leftRotate(node.left);
            }
            node = rightRotate(node);
        }

        if (balanceFactor(node) < -1) {
            if (balanceFactor(node.right) > 0) {
                node.right = rightRotate(node.right);
            }
            node = leftRotate(node);
        }

        return node;
    }

    public void insert(T data) {
        root = insertRecursive(root, data);
    }

    private Node<T> insertRecursive(Node<T> node, T data) {
        if (node == null) {
            size++;
            return new Node<T>(data);
        }

        int comparison = data.compareTo(node.data);

        if (comparison < 0) {
            node.left = insertRecursive(node.left, data);
            node.left.parent = node;
        } else if (comparison > 0) {
            node.right = insertRecursive(node.right, data);
            node.right.parent = node;
        } else {
            return node;
        }

        node.updateHeight();

        return rebalanceNode(node);
    }

    public boolean contains(T data) {
        Node<T> currentNode = root;

        while (currentNode != null) {
            int comparison = data.compareTo(currentNode.data);

            if (comparison < 0) {
                currentNode = currentNode.left;
            } else if (comparison > 0) {
                currentNode = currentNode.right;
            } else {
                return true;
            }
        }

        return false;
    }

    public void remove(T data) {
        root = removeRecursive(root, data);
    }

    private Node<T> removeRecursive(Node<T> node, T data) {
        if (node == null) {
            return null;
        }

        int comparison = data.compareTo(node.data);

        if (comparison < 0) {
            node.left = removeRecursive(node.left, data);
        } else if (comparison > 0) {
            node.right = removeRecursive(node.right, data);
        } else {
            size--;
            Node<T> parent = node.parent;
            boolean wasLeft = (parent != null && parent.left == node);
            if (node.left != null && node.right != null) {
                node = removeTwoChild(node);
            } else if (node.left != null || node.right != null) {
                node = removeOneChild(node);
            } else {
                node = removeLeaf(node);
            }

            if (parent != null) {
                if (wasLeft) {
                    parent.left = node;
                } else {
                    parent.right = node;
                }
            }
        }

        if (node == null) {
            return null;
        }
        node.updateHeight();

        return rebalanceNode(node);
    }

    private Node<T> removeLeaf(Node<T> node) {
        if (node.parent == null) {
            root = null;
        } else if (node == node.parent.left) {
            node.parent.left = null;
        } else {
            node.parent.right = null;
        }

        return null;
    }

    private Node<T> removeOneChild(Node<T> node) {
        if (node == root) {
            if (root.left != null) {
                root = root.left;
            } else {
                root = root.right;
            }
        } else {
            if (node == node.parent.left) {
                if (node.left != null) {
                    node.parent.left = node.left;
                    node = node.left;
                } else {
                    node.parent.left = node.right;
                    node = node.right;
                }
            } else {
                if (node.left != null) {
                    node.parent.right = node.left;
                    node = node.left;
                } else {
                    node.parent.right = node.right;
                    node = node.right;
                }
            }
        }

        return node;
    }

    private Node<T> removeTwoChild(Node<T> node) {
        Node<T> predecessor = node.left;

        while (predecessor.right != null) {
            predecessor = predecessor.right;
        }

        node.data = predecessor.data;

        if (predecessor.left != null || predecessor.right != null) {
            predecessor = removeOneChild(predecessor);
        } else {
            predecessor = removeLeaf(predecessor);
        }

        return node;
    }

    private String inOrderTraverse(Node<T> node) {
        if (node == null) {
            return "";
        }

        StringBuilder result = new StringBuilder();

        result.append(inOrderTraverse(node.left)).append(node.data).append(" ")
                .append(inOrderTraverse(node.right));

        return result.toString();
    }

    public String inOrder() {
        return inOrderTraverse(root);
    }

    public String toString() {
        return inOrder();
    }

    private String preOrderTraverse(Node<T> node) {
        if (node == null) {
            return "";
        }

        StringBuilder result = new StringBuilder();

        result.append(node.data).append(" ").append(preOrderTraverse(node.left))
                .append(preOrderTraverse(node.right));

        return result.toString();
    }

    public String preOrder() {
        return preOrderTraverse(root);
    }

    private String postOrderTraverse(Node<T> node) {
        if (node == null) {
            return "";
        }

        StringBuilder result = new StringBuilder();

        result.append(postOrderTraverse(node.left))
                .append(postOrderTraverse(node.right)).append(node.data).append(" ");

        return result.toString();
    }

    public String postOrder() {
        return postOrderTraverse(root);
    }

    public String levelOrder() {
        if (isEmpty()) {
            return "";
        }

        ArrayDeque<Node<T>> queue = new ArrayDeque<>();
        StringBuilder result = new StringBuilder();

        queue.add(root);

        while (!queue.isEmpty()) {
            Node<T> node = queue.pollFirst();

            result.append(node.data).append(" ");

            if (node.left != null) {
                queue.add(node.left);
            }

            if (node.right != null) {
                queue.add(node.right);
            }
        }

        return result.toString();
    }

    public void clear() {
        root = null;
        size = 0;
    }

    public static void main(String[] args) {
        AVLTree<Integer> myAVL = new AVLTree<>();

        for (int i = 0; i < 15; ++i) {
            System.out.println("Inserting " + (17 * (i + 1) + 3) % 31);
            myAVL.insert((17 * (i + 1) + 3) % 31);
        }

        System.out.println("Pre-order traversal: " + myAVL.preOrder());
        System.out.println("In-order traversal: " + myAVL);
        System.out.println("Post-order traversal: " + myAVL.postOrder());
        System.out.println("Level-order traversal: " + myAVL.levelOrder());
        System.out.println("Size: " + myAVL.size());
        System.out.println("Empty? " + myAVL.isEmpty());

        System.out.println("\nContains 7? " + myAVL.contains(7));
        System.out.println("\nContains 11? " + myAVL.contains(11));

        System.out.println("\nTesting removal of leaf (18).");
        myAVL.remove(18);
        System.out.println("Level-order traversal: " + myAVL.levelOrder());

        System.out.println("\nTesting removal of node with one child (6).");
        myAVL.remove(6);
        System.out.println("Level-order traversal: " + myAVL.levelOrder());

        System.out.println("\nTesting removal of node with two children (20).");
        myAVL.remove(20);
        System.out.println("Level-order traversal: " + myAVL.levelOrder());

        System.out.println("\nClearing tree. The traversals should be blank below this line.");
        myAVL.clear();

        System.out.println("Pre-order traversal: " + myAVL.preOrder());
        System.out.println("In-order traversal: " + myAVL);
        System.out.println("Post-order traversal: " + myAVL.postOrder());
        System.out.println("Level-order traversal: " + myAVL.levelOrder());
        System.out.println("Size: " + myAVL.size());
        System.out.println("Empty? " + myAVL.isEmpty());
    }
}

File: avl_tree/avl_tree.c
#include "avl_tree.h"

avl_t *create_avl(int (*comparator)(void *, void *),
                  void (*printer)(void *),
                  void (*deleter)(void *))
{
    avl_t *tree = malloc(sizeof(avl_t));

    if( !tree )
    {
        return nullptr;
    }

    tree->root = nullptr;
    tree->size = 0;
    tree->comparator = comparator;
    tree->printer = printer;
    tree->deleter = deleter;

    return tree;
}

node_t *create_node(void *data)
{
    node_t *node = malloc(sizeof(node_t));

    if( !node )
    {
        return nullptr;
    }

    node->data = data;
    node->parent = node->left = node->right = nullptr;
    node->height = 0;

    return node;
}

node_t *find_node(avl_t *tree, void *data)
{
    if( !tree )
    {
        return nullptr;
    }

    node_t *current_node = tree->root;

    while( current_node )
    {
        int comparison = tree->comparator(data, current_node->data);

        if( comparison < 0 )
        {
            current_node = current_node->left;
        }
        else if( comparison > 0 )
        {
            current_node = current_node->right;
        }
        else
        {
            return current_node;
        }
    }

    return nullptr;
}

int node_height(node_t *node)
{
    return (node) ? node->height : -1;
}

void update_height(node_t *node)
{
    if( !node )
    {
        return;
    }

    node->height = 1 + MAX(node_height(node->left), node_height(node->right));
}

int balance_factor(node_t *node)
{
    if( !node )
    {
        return 0;
    }

    return node_height(node->left) - node_height(node->right);
}

node_t *left_rotate(node_t *node)
{
    if( !node->right )
    {
        return nullptr;
    }

    node_t *right = node->right;
    node->right = right->left;

    right->left = node;
    right->parent = node->parent;

    if( right->parent )
    {
        if( node->parent->left == node )
        {
            node->parent->left = right;
        }
        else
        {
            node->parent->right = right;
        }
    }

    node->parent = right;

    update_height(node);
    update_height(right);

    return right;
}

node_t *right_rotate(node_t *node)
{
    if( !node->left )
    {
        return nullptr;
    }

    node_t *left = node->left;
    node->left = left->right;

    left->right = node;
    left->parent = node->parent;

    if( left->parent )
    {
        if( node->parent->left == node )
        {
            node->parent->left = left;
        }
        else
        {
            node->parent->right = left;
        }
    }

    node->parent = left;

    update_height(node);
    update_height(left);

    return left;
}

node_t *rebalance_node(node_t *node)
{
    if( balance_factor(node) > 1 )
    {
        if( balance_factor(node->left) < 0 )
        {
            node->left = left_rotate(node->left);
        }

        node = right_rotate(node);
    }

    if( balance_factor(node) < -1 )
    {
        if( balance_factor(node->right) > 0 )
        {
            node->right = right_rotate(node->right);
        }

        node = left_rotate(node);
    }

    return node;
}

void insert_node(avl_t *tree, void *data)
{
    tree->root = insert_recursive(tree, tree->root, data);
}

node_t *insert_recursive(avl_t *tree, node_t *node, void *data)
{
    if( !node )
    {
        ++tree->size;
        return create_node(data);
    }

    int comparison = tree->comparator(data, node->data);

    if( comparison < 0 )
    {
        node->left = insert_recursive(tree, node->left, data);
        node->left->parent = node;
    }
    else if( comparison > 0 )
    {
        node->right = insert_recursive(tree, node->right, data);
        node->right->parent = node;
    }
    else
    {
        return node;
    }

    update_height(node);

    return rebalance_node(node);
}

bool contains(avl_t *tree, void *data)
{
    return find_node(tree, data);
}

void *remove_node(avl_t *tree, void *data)
{
    if( !tree || !data )
    {
        return nullptr;
    }

    node_t *removed_node = nullptr;
    tree->root = remove_recursive(tree, tree->root, &removed_node, data);

    if( !removed_node )
    {
        return nullptr;
    }

    data = removed_node->data;
    free(removed_node);

    return data;
}

node_t *
remove_recursive(avl_t *tree, node_t *node, node_t **removed_node, void *data)
{
    if( !tree || !node || !data )
    {
        return nullptr;
    }

    int comparison = tree->comparator(data, node->data);

    if( comparison < 0 )
    {
        node->left = remove_recursive(tree, node->left, removed_node, data);
    }
    else if( comparison > 0 )
    {
        node->right = remove_recursive(tree, node->right, removed_node, data);
    }
    else
    {
        --tree->size;

        node_t *parent = node->parent;
        bool was_left = (parent && parent->left == node) ? true : false;
        bool was_leaf = false;

        if( node->left && node->right )
        {
            node = remove_two_child(tree, node, removed_node);
        }
        else if( node->left || node->right )
        {
            node = remove_one_child(tree, node, removed_node);
        }
        else
        {
            node = remove_leaf(tree, node, removed_node);
            was_leaf = true;
        }

        node->parent = parent;

        if( parent )
        {
            if( was_left )
            {
                parent->left = node;
            }
            else
            {
                parent->right = node;
            }
        }

        if( was_leaf )
        {
            update_height(parent);
            rebalance_node(parent);
            return nullptr;
        }
    }

    update_height(node);

    return rebalance_node(node);
}

node_t *remove_leaf(avl_t *tree, node_t *node, node_t **removed_node)
{
    if( is_empty(tree) || !node )
    {
        return nullptr;
    }

    if( node == tree->root )
    {
        tree->root = nullptr;
    }
    else if( node->parent->left == node )
    {
        node->parent->left = nullptr;
    }
    else
    {
        node->parent->right = nullptr;
    }

    *removed_node = node;

    return node;
}

node_t *remove_one_child(avl_t *tree, node_t *node, node_t **removed_node)
{
    if( is_empty(tree) || !node || (node->left && node->right) )
    {
        return nullptr;
    }

    node_t **relink_node;
    if( node == tree->root )
    {
        relink_node = &tree->root;
    }
    else
    {
        relink_node = (node == node->parent->left) ? &node->parent->left
                                                   : &node->parent->right;
    }

    *removed_node = node;

    node = *relink_node = (node->left) ? node->left : node->right;

    if( node->parent->parent )
    {
        node->parent = node->parent->parent;
    }
    else
    {
        node->parent = nullptr;
    }

    return node;
}

node_t *remove_two_child(avl_t *tree, node_t *node, node_t **removed_node)
{
    if( is_empty(tree) || !node )
    {
        return nullptr;
    }

    node_t *predecessor = node->left;

    while( predecessor->right )
    {
        predecessor = predecessor->right;
    }

    void *data = node->data;
    node->data = predecessor->data;

    if( predecessor->left || predecessor->right )
    {
        remove_one_child(tree, predecessor, removed_node);
    }
    else
    {
        remove_leaf(tree, predecessor, removed_node);
    }

    (*removed_node)->data = data;

    return node;
}

void clear(avl_t *tree)
{
    if( !tree )
    {
        return;
    }

    clear_recursive(tree->root, tree->deleter);
    tree->root = nullptr;
    tree->size = 0;
}

void clear_recursive(node_t *node, void (*deleter)(void *))
{
    if( !node )
    {
        return;
    }

    clear_recursive(node->left, deleter);
    clear_recursive(node->right, deleter);

    deleter(node->data);
    free(node);
}

void pre_order_print(avl_t *tree)
{
    if( !tree )
    {
        return;
    }

    pre_order_recurse(tree->root, tree->printer);
    printf("\n\n");
}

void pre_order_traverse(avl_t *tree, void (*func)(void *))
{
    if( !tree )
    {
        return;
    }

    pre_order_recurse(tree->root, func);
}

void pre_order_recurse(node_t *node, void (*func)(void *))
{
    if( !node )
    {
        return;
    }

    func(node->data);
    pre_order_recurse(node->left, func);
    pre_order_recurse(node->right, func);
}

void in_order_print(avl_t *tree)
{
    if( !tree )
    {
        return;
    }

    in_order_recurse(tree->root, tree->printer);
    printf("\n\n");
}

void in_order_traverse(avl_t *tree, void (*func)(void *))
{
    if( !tree )
    {
        return;
    }

    in_order_recurse(tree->root, func);
}

void in_order_recurse(node_t *node, void (*func)(void *))
{
    if( !node )
    {
        return;
    }

    in_order_recurse(node->left, func);
    func(node->data);
    in_order_recurse(node->right, func);
}

void post_order_print(avl_t *tree)
{
    if( !tree )
    {
        return;
    }

    post_order_recurse(tree->root, tree->printer);
    printf("\n\n");
}

void post_order_traverse(avl_t *tree, void (*func)(void *))
{
    if( !tree )
    {
        return;
    }

    post_order_recurse(tree->root, func);
}

void post_order_recurse(node_t *node, void (*func)(void *))
{
    if( !node )
    {
        return;
    }

    post_order_recurse(node->left, func);
    post_order_recurse(node->right, func);
    func(node->data);
}

void level_order_print(avl_t *tree)
{
    if( !tree )
    {
        return;
    }

    level_order_traverse(tree, tree->printer);
    printf("\n\n");
}

void level_order_traverse(avl_t *tree, void (*func)(void *))
{
    if( !tree || !tree->root )
    {
        return;
    }

    deque_t *queue = CREATE_DEQUE(node_t *);

    push_back(queue, &tree->root);

    while( queue->size > 0 )
    {
        node_t *node;
        pop_front(queue, &node);

        if( !node )
        {
            continue;
        }

        func(node->data);

        if( node->left )
        {
            push_back(queue, &(node->left));
        }

        if( node->right )
        {
            push_back(queue, &node->right);
        }
    }

    delete_deque(&queue);
}

bool is_empty(avl_t *tree)
{
    return !tree || tree->size == 0;
}

void delete_avl(avl_t **tree)
{
    if( !tree || !*tree )
    {
        return;
    }

    clear(*tree);
    free(*tree);
    *tree = nullptr;
}
